source/extensions/filters/http/header_mutation/query_params_evaluator.cc

```diff
@@ -0,0 +1,76 @@
+#include "source/extensions/filters/http/header_mutation/query_params_evaluator.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "envoy/http/query_params.h"
+#include "envoy/stream_info/stream_info.h"
+
+#include "source/common/formatter/substitution_formatter.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace HeaderMutation {
+
+using Http::Utility::QueryParamsMulti;
+
+QueryParamsEvaluator::QueryParamsEvaluator(
+    const Protobuf::RepeatedPtrField<KeyValueMutationProto>& query_param_mutations)
+    : formatter_(std::make_unique<Formatter::FormatterImpl>("", true)) {
+
+  for (const auto& query_param : query_param_mutations) {
+    mutations_.emplace_back(query_param);
+  }
+}
+
+void QueryParamsEvaluator::evaluateQueryParams(Http::RequestHeaderMap& headers,
+                                               const StreamInfo::StreamInfo& stream_info) const {
+  if (mutations_.empty()) {
+    return;
+  }
+
+  absl::string_view path = headers.getPathValue();
+  QueryParamsMulti query_params = QueryParamsMulti::parseAndDecodeQueryString(path);
+
+  Formatter::HttpFormatterContext ctx{&headers};
+  for (const auto& mutation : mutations_) {
+    if (!mutation.remove().empty()) {
+      query_params.remove(mutation.remove());
+    } else {
+      const auto value_option = mutation.append();
+      const auto key = value_option.entry().key();
+      const auto value = value_option.entry().value();
+      const auto formatter = std::make_unique<Formatter::FormatterImpl>(value, true);
```

Comment:
Please initialize the formatter when loading configuration to avoid do this at runtime.

Reply:
The formatter needs the `value`?

Reply:
Yeah, you need to parse the value when you constructing `QueryParamsEvaluator`. Creating the formatter has big overhead and my throw exception. It's not allowed to do that at key data path.

Reply:
I think this is problem is still not be resolved.

We should load/create the formatter when loading configuration.

You can take the header mutations implemention as an example.

/lgtm api 

/wait